File: SRP/misconceptions/Order.ts
export class Order {
    size: number = 2;
}

export class OrderValidator {
    isValid(order: Order): boolean {
        return order.size > 0;
    }
}

export class OrderSaver {
    save(order: Order): void {
        console.log("Saving order to DB");
    }
}

export class OrderEmailSender {
    send(order: Order): void {
        console.log("Sending confirmation email");
    }
}

export class OrderLogger {
    log(order: Order): void {
        console.log("Logging order");
    }
}

export class OrderService {
    constructor(
        private readonly validator: OrderValidator,
        private readonly saver: OrderSaver,
        private readonly emailSender: OrderEmailSender,
        private readonly logger: OrderLogger
    ) { }

    placeOrder(order: Order): void {
        if (!this.validator.isValid(order)) {
            throw new Error("Invalid order");
        }

        this.saver.save(order);
        this.emailSender.send(order);
        this.logger.log(order);
    }
}

File: SRP/misconceptions/Order.java
package SRP;

public class Order {
    int size = 2;
}

class OrderValidator {
    boolean isValid(Order order) {
        return order.size > 0;
    }
}

class OrderSaver {
    void save(Order order) {
        System.out.println("Saving order to DB");
    }
}

class OrderEmailSender {
    void send(Order order) {
        System.out.println("Sending confirmation email");
    }
}

class OrderLogger {
    void log(Order order) {
        System.out.println("Logging order");
    }
}

class OrderService {
    void placeOrder(Order order) {
        validate(order);
        persist(order);
        notifyUser(order);
        audit(order);
    }

    private void audit(Order order) {
        throw new UnsupportedOperationException("Unimplemented method 'audit'");
    }

    private void notifyUser(Order order) {
        throw new UnsupportedOperationException("Unimplemented method 'notifyUser'");
    }

    private void persist(Order order) {
        throw new UnsupportedOperationException("Unimplemented method 'persist'");
    }

    private void validate(Order order) {
        throw new UnsupportedOperationException("Unimplemented method 'validate'");
    }
}
